dts: arm: silabs: Restructure Series 2 DTS content

asmellby opened the pull request:
.dtsi files for Silicon Labs SoCs are mixed together in a single directory covering 5 generations of SoCs with 3 completely different lineages. Within Series 2, naming is also inconsistent, and .dtsi files are not separated in a way that makes them reusable or composable.

This PR cleans up the Series 2 .dtsi files by structuring them as follows:

* One directory per generic family (`xg22/` = series 2 config 2)
* One base .dtsi file describing the MCU part of the device (`xg22.dtsi` -- shareable across pg22 MCUs and bg22/fg22/mg22 RF MCUs)
* One .dtsi file adding radio capability to the base file (`efr32xg22.dtsi` -- shareable between bg22/fg22/mg22 RF MCUs)
* If not all radio capable devices support all RF protocols, one .dtsi file per device family (`ef32bg22.dtsi`)
* SoC specific .dtsi files with specific peripheral capabilities, flash and ram size, etc

Boards continue to include the SoC-specific .dtsi file.

While each generic family modified in this PR currently only contains a single series, there are pending PRs for adding
* EFM32PG23, which would benefit from a shared xg23.dtsi file https://github.com/zephyrproject-rtos/hal_silabs/pull/95
* EFM32PG28 and EFR32ZG28 by separate contributors, which would benefit from a structure allowing a shared xg28.dtsi file https://github.com/zephyrproject-rtos/zephyr/pull/89597

asmellby committed:
dts: arm: silabs: Move efr32mg21 to xg21 directory

Introduce subdirectory for xg21 socs.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
dts: arm: silabs: Move efr32bg22 to xg22 directory

Introduce subdirectory for xg22 socs.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
dts: arm: silabs: Move efr32zg23 to xg23 directory

Introduce subdirectory for xg23 socs.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
dts: arm: silabs: Move efr32mg24 to xg24 directory

Introduce subdirectory for xg24 socs.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
dts: arm: silabs: Move efr32bg27 to xg27 directory

Introduce subdirectory for xg27 socs.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

jerome-pouiller commented:
Don't you think we should introduce `efr32xg21.dtsi`, `efr32mg22.dtsi`, etc... Even if these files only include the parent .dtsi, it would help to keep things consistent?

asmellby committed:
dts: arm: silabs: Refactor xg29 directory

Align xg29 directory with other Series 2 families, introducing
separate .dtsi files per device family inheriting the common
xg29.dtsi file.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby commented:
> Don't you think we should introduce `efr32xg21.dtsi`, `efr32mg22.dtsi`, etc... Even if these files only include the parent .dtsi, it would help to keep things consistent?

Works for me.

yishai1999 commented:
Big like

asmellby commented on dts/arm/silabs/xg24/efr32mg24.dtsi:
```diff
@@ -0,0 +1,7 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <silabs/xg24/efr32xg24.dtsi>
```
Identical for now. Once 802.15.4 support is implemented, it will only be in efr32mg24.dtsi and not in efr32bg24.dtsi.

jerome-pouiller commented:
Should we place somewhere in the tree the explanations you provide in the PR description? Maybe in the same vein than `boards/silabs/index.rst`?

kartben reviewed (changes requested):
probably needs a mention in migration guide since out-of-tree boards will be impacted

asmellby committed:
doc: migration-guide-4.2: Silabs SoC DTS restructure

Describe changes users need to do in their out of tree board definitions
after the SoC DTS files have been moved into subdirectories.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>
